text/0431-sagemaker-l2-endpoint.md

```diff
@@ -0,0 +1,1150 @@
+production faster with much less effort and lower cost.
+
+## Installation
+
+Install the module:
+
+```console
+$ npm i @aws-cdk/aws-sagemaker
+```
+
+Import it into your code:
+
+```typescript
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+```
+
+## Model
+
+By creating a model, you tell Amazon SageMaker where it can find the model components. This includes
+the S3 path where the model artifacts are stored and the Docker registry path for the image that
+contains the inference code. The `ContainerDefinition` interface encapsulates both the specification
+of model inference code as a `ContainerImage` and an optional set of artifacts as `ModelData`.
+
+### Container Images
+
+Inference code can be stored in the Amazon EC2 Container Registry (Amazon ECR), which is specified
+via `ContainerDefinition`'s `image` property which accepts a class that extends the `ContainerImage`
+abstract base class.
+
+#### `EcrImage`
+
+Reference an image available within ECR:
+
+```typescript
+import * as ecr from '@aws-cdk/aws-ecr';
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+const repository = ecr.Repository.fromRepositoryName(this, 'Repository', 'repo');
+const image = sagemaker.ContainerImage.fromEcrRepository(repository, 'tag');
+```
+
+#### `AssetImage`
+
+Reference a local directory containing a Dockerfile:
+
+```typescript
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+import * as path from 'path';
+
+const image = sagemaker.ContainerImage.fromAsset(this, 'Image', {
+  directory: path.join('path', 'to', 'Dockerfile', 'directory')
+});
+```
+
+### Model Artifacts
+
+Models are often associated with model artifacts, which are specified via the `modelData` property
+which accepts a class that extends the `ModelData` abstract base class. The default is to have no
+model artifacts associated with a model.
+
+#### `S3ModelData`
+
+Reference an S3 bucket and object key as the artifacts for a model:
+
+```typescript
+import * as s3 from '@aws-cdk/aws-s3';
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+const bucket = new s3.Bucket(this, 'MyBucket');
+const modelData = sagemaker.ModelData.fromBucket(bucket, 'path/to/artifact/file.tar.gz');
+```
+
+#### `AssetModelData`
+
+Reference local model data:
+
+```typescript
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+import * as path from 'path';
+
+const modelData = sagemaker.ModelData.fromAsset(this, 'ModelData',
+  path.join('path', 'to', 'artifact', 'file.tar.gz'));
+```
+
+### `Model`
+
+The `Model` construct associates container images with their optional model data.
+
+#### Single Container Model
+
+In the event that a single container is sufficient for your inference use-case, you can define a
+single-container model:
+
+```typescript fixture=with-assets
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+const model = new sagemaker.Model(this, 'PrimaryContainerModel', {
+  container: {
+    image: image,
+    modelData: modelData,
+  }
```

Comment:
We prefer a [flat namespace whenever possible](https://github.com/aws/aws-cdk/blob/master/docs/DESIGN_GUIDELINES.md#flat). Does the additional nesting of `container` here provide a lot of value?

Reply:
Would it make sense for `Container` to be a class?